File: leetcode/Decode_Ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        if(s.empty()) return 0;
        int n = s.size();
        
        vector<int> t(n+1, 0);
        t[0] = 1;
        
        for(int i = 1; i <= n; i++){
            if(s[i-1] != '0') t[i] += t[i-1];
            if(isValid(s, i-2)) t[i] += t[i-2];
        }
        
        return t[n];
    }
private:
    int isValid(string &s, int i){
        if(i < 0) return false;
        
        int x = stoi(s.substr(i, 2));
        if(x >= 10 && x <= 26) return true;
        else return false;
    }
};

File: leetcode/java/DecodeWays.java
class Solution {
    public int numDecodings(String s) {
        if(s == null || s.length() == 0) return 0;
        int n = s.length();
        int[] t = new int[n+1];
        t[0] = 1;
        
        for(int i = 1; i <= n; i++) {
            if(s.charAt(i-1) != '0') t[i] += t[i-1];
            if(isRange(s, i-2, i)) t[i] += t[i-2];
        }
        return t[n];
    }
    
    private boolean isRange(String st, int s, int e) {
        if(s < 0) return false;
        
        int base = 1;
        int result = 0;
        for(int i = s; i < e; i++) {
            result = result*base + st.charAt(i)-'0';
            base *= 10;
        }
        return result >= 10 && result <= 26;
    }
}
